src/main/scala/components/memory-async.scala

```diff
@@ -0,0 +1,378 @@
+        val offset = storedWrite.address (1, 0)
+        val readdata = Wire (UInt (32.W))
+        readdata := io.response.bits
+        val data = Wire (UInt (32.W))
+        // Mask the portion of the existing data so it can be or'd with the writedata
+        when (storedWrite.maskmode === 0.U) {
+          data := io.response.bits & ~(0xff.U << (offset * 8.U))
+        } .otherwise {
+          data := io.response.bits & ~(0xffff.U << (offset * 8.U))
+        }
+        writedata := data | (storedWrite.writedata << (offset * 8.U))
+      } .otherwise {
+        // Write the entire word
+        writedata := storedWrite.writedata 
+      }
+
+      // Program the memory to issue a write
+      val request = Wire (new Request)
+      request.address   := storedWrite.address
+      request.writedata := writedata
+      request.operation := 2.U
+      io.request.bits  := request
+      io.request.valid := true.B
+
+      // Wipe the stored value as it is no longer needed
+      storedWrite := 0.U.asTypeOf(new PartialWrite)
+    } .otherwise {
+      // Perform masking and sign extension on read data when memory is outputting it
+      val readdata_mask      = Wire(UInt(32.W))
+      val readdata_mask_sext = Wire(UInt(32.W))
+
+      when (io.maskmode === 0.U) {
+        // Byte
+        readdata_mask := io.response.bits & 0xff.U
+      } .elsewhen (io.maskmode === 1.U) {
+        // Half-word
+        readdata_mask := io.response.bits & 0xffff.U
+      } .otherwise {
+        // Word
+        readdata_mask := io.response.bits
+      }
+
+      when (io.sext) {
+        when (io.maskmode === 0.U) {
+          readdata_mask_sext := Cat(Fill(24, readdata_mask(7)),  readdata_mask(7, 0))
+        } .elsewhen (io.maskmode === 1.U) {
+          readdata_mask_sext := Cat(Fill(26, readdata_mask(15)), readdata_mask(15,0))
+        } .otherwise {
+          readdata_mask_sext := readdata_mask
+        }
+      } .otherwise {
+        readdata_mask_sext := readdata_mask
+      }
+
+      io.readdata := readdata_mask_sext
+    }
+  }
+}
+
+
+/**
+ * The modified asynchronous form of the dual ported memory module. 
+ * When io.imem.request.valid or io.imem.request.valid is true and the memory is ready for an operation,
+ * this memory module simulates the latency of real DRAM by pushing memory accesses into pipes that delay 
+ * the request for a configurable latency.
+ *
+ * As with the synced memory module, this memory should only be instantiated in the Top file,
+ * and never within the actual CPU.
+ *
+ * The I/O for this module is defined in [[AsyncMemIO]].
+ */
+class DualPortedAsyncMemory(size: Int, memfile: String, latency: Int) extends Module {
+  val io = IO(new Bundle {
+    val imem = new AsyncMemIO
+    val dmem = new AsyncMemIO
+  })
+  io := DontCare
+
+  assert(latency > 0) // Check for attempt to make combinational memory
+
+  val memory    = Mem(math.ceil(size.toDouble/4).toInt, UInt(32.W))
```

Comment:
We could (at some point) change the `32.U` to be a parameter to set different block sizes. Future commit, though :)